### PR TITLE
[AArch64] Improve code generation of bool vector reduce operations

### PR DESCRIPTION
**Pull request by Il-Capitano**
* Avoid unnecessary truncation of comparison results in vecreduce_xor
* Optimize generated code for vecreduce_and and vecreduce_or by comparing against 0.0 to check if all/any of the values are set

Alive2 proof of vecreduce_and and vecreduce_or transformation: https://alive2.llvm.org/ce/z/SRfPtw

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:15841]: `DAG.getConstant(-1, DL, VT)`. 
Although maybe `DAG.getNOT` instead.

Comment (Il-Capitano) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:15841]: > `DAG.getConstant(-1, DL, VT)`.

This is what I tried initially, but that leads to bad code generation in some cases, e.g. in test/CodeGen/reduce-and.ll this is part of the diff compared to the current version:
```patch
@@ -42,7 +42,8 @@ define i1 @test_redand_v2i1(<2 x i1> %a) {
 define i1 @test_redand_v4i1(<4 x i1> %a) {
 ; CHECK-LABEL: test_redand_v4i1:
 ; CHECK:       // %bb.0:
-; CHECK-NEXT:    mvn v0.8b, v0.8b
+; CHECK-NEXT:    movi v1.4h, #1
+; CHECK-NEXT:    eor v0.8b, v0.8b, v1.8b
 ; CHECK-NEXT:    shl v0.4h, v0.4h, #15
 ; CHECK-NEXT:    cmlt v0.4h, v0.4h, #0
 ; CHECK-NEXT:    fcmp d0, #0.0
```

I quickly checked, and the same thing happens when using `DAG.getNOT`.

I looked into this previously, and found two separate changes that fixed codegen when using `DAG.getConstant(-1, DL, VT)`:
1. The legalization logic for `SPLAT_VECTOR` is different from `BUILD_VECTOR` in the way they handle promotion of boolean operands. In `SPLAT_VECTOR` a `Constant::i1<-1>` gets zero-extended to `Constant::i32<1>`, while in `BUILD_VECTOR` it gets sign-extended to `Constant::i32<-1>`. Copying the logic used in `BUILD_VECTOR` legalization over to `SPLAT_VECTOR` legalization fixes the issue.
2. If we pass `AllowTruncation=true` to the `isConstOrConstSplat` call in [CodeGen/SelectionDAG/TargetLowering.cpp:1617](https://github.com/llvm/llvm-project/blob/469ac118418fff2fc07e5705ff527405060ac586/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L1617) and handle the constant value appropriately, it also eliminates the issue.

When using `DAG.getNOT`, only the second option would fix it, because it uses `Constant::i32<1>` values in the generated node.

I'm not sure what solution would be preferred. The two changes above are unrelated to this PR, so I could open separate PRs for them, and make this one depend on those, however I couldn't come up with a test case for either that showed their effects.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (Il-Capitano) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:15862]: I should use `DAG.getSetCC` here.

**Review by lawben — COMMENTED** [review]
LGMT. Using the floating point compare seems like a neat idea. But you should probably wait for the approval of another reviewer, as I'm not sure if I'm missing an edge case or not.

**Review by davemgreen — APPROVED** [review]
Thanks for the explanation. I hadn't realized it was an i1 not.

LGTM, thanks.

File: llvm/test/CodeGen/AArch64/reduce-and.ll
Comment (david-arm) [inline llvm/test/CodeGen/AArch64/reduce-and.ll:26]: I think the lowering is quite clever here, but is there now an issue with serialisation if you have multiple reductions? Suppose your IR looks like this:

```
  %or_result1 = call i1 @llvm.vector.reduce.and.v2i1(<2 x i1> %a)
  %or_result2 = call i1 @llvm.vector.reduce.and.v2i1(<2 x i1> %b)
  %or_result = or i1 %or_result1, %or_result2
  ret i1 %or_result
```

I haven't checked the code with and without this patch, but I imagine previously we could quite happily have interleaved the instructions like this:

```
  uminp v0 ...
  uminp v1 ...
  fmov w8, s0
  fmov w9, s1
  and w0, w8, 0x1
  and w1, w9, 0x1
  or w0, w0, w1
```

whereas now due to the single CC register we have to serialise:

```
  fcmp d0, #0.0
  cset w0, eq
  fcmp d1, #0.0
  cset w1, eq
  or w0, w0, w1
...
```

However, I can see how this new version is efficient if the result is then used for control flow:

```
  %or_result = call i1 @llvm.vector.reduce.and.v2i1(<2 x i1> %a)
  br i1 %or_result, ...
```

Do you have any examples showing where this patch helps improve performance?

Comment (Il-Capitano) [reply llvm/test/CodeGen/AArch64/reduce-and.ll:26]: For performance differences I originally tested a simple loop similar to this:
```cpp
void test(bool *dest, float32x4_t *p, std::size_t n) {
    for (std::size_t i = 0; i < n; ++i) {
        dest[i] = __builtin_reduce_or(p[i] < 0.0);
    }
}
```
My change typically resulted in a 10-15% improvement on various CPUs.
My original motivating use case was something like this:
```cpp
float32x4_t x = ...;
// ...
x = f(x);
if (__builtin_reduce_or(x < 0.0)) return;
x = g(x);
if (__builtin_reduce_or(x < 0.0)) return;
// ...
```
which should benefit a bit more from this change, since the reduction result is used for control flow.

For the case of `or(reduce_and(x), reduce_and(y))`, you make a good point. Currently LLVM generates [this](https://godbolt.org/z/n5KKvnTrh):
```
        uminv   b0, v0.8b
        uminv   b1, v1.8b
        fmov    w8, s0
        fmov    w9, s1
        orr     w8, w8, w9
        and     w0, w8, #0x1
        ret
```
With my change LLVM generates something like this:
```
        fcmp d0, #0.0
        cset w8, eq
        fcmp d1, #0.0
        csinc w0, w8, wzr, ne
        ret
```
In this example both snippets have a max dependency length of 4, but if we were to `or` together more `reduce_and` operations, the generated code would get worse after this change, which can be a concern. Although I'm not sure how common such a pattern is in real code.

Similar patterns like `and(reduce_and(x), reduce_and(y))` get folded to `reduce_and(and(x, y))` by instcombine, so that shouldn't be an issue.

I guess the question is how common the pattern you bring up is in real-world code, and does the potential regression in that case outweigh the improvements in other cases. I don't really have a good answer for this though.

Comment (david-arm) [reply llvm/test/CodeGen/AArch64/reduce-and.ll:26]: Thanks for explaining! And just to be clear I'm not saying this will definitely be an issue and I won't hold up the patch for it. I was just curious to see the motivating examples, that's all.

**Comment by Il-Capitano** [comment]
Are there any other thoughts on this, or can we go ahead with the merge?

**Comment by davemgreen** [comment]
Thanks!

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (nikic) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:15863]: FYI there is a `DAG.getAllOnesConstant(DL, MVT::i32)` helper for this.

**Comment by Il-Capitano** [comment]
@davemgreen @nikic 
It looks like this breaks some things when compiling with `-Ofast`: https://linaro.atlassian.net/browse/LLVM-1481

I was able to reproduce the issue with this test case: https://github.com/fujitsu/compiler-test-suite/blob/main/C/0030/0030_0066.c

The issue seems to be triggered during LTO. If I compile the test case to assembly, there's no difference between `-O3` and `-Ofast`.

**Comment by davemgreen** [comment]
I see, with fast math.. I'll revert this for now if that is OK.

I think it might be conflicting with DAZ/FTZ that gets set by fast-math. The start-up object sets the flags in FPCR.

**Comment by Il-Capitano** [comment]
You're right, the issue is that the fast-math start-up object sets the FZ bit (bit 24) of FPCR, meaning that [_denormalized single-precision and double-precision inputs to, and outputs from, floating-point instructions are flushed to zero_](https://developer.arm.com/documentation/ddi0601/2024-09/AArch64-Registers/FPCR--Floating-point-Control-Register). This means that we can't reliably use `fcmp d0, #0.0` to check for an all-zero bit pattern.

To me it seems there's no way to use the `fcmp` trick and get correct lowering in all cases, since the setting of FPCR is configured at link time, i.e. a TU compiled with `-O3` can be affected if the `-mdaz-ftz` flag is used in the linking command.

There's one part of this patch that could still be used to optimize boolean reductions (used for `reduce_xor` here). I'll open a new PR for that in the next few days.

**Pull request merged**
